DoSomethingArchive/legacy-website
# Campaign Action Page Gallery Refactoring

Fixes #5980
Fixes #5170
Refs #5874 (partially fixes `total` for RB items; need to double check for campaigns)
#### What's this PR do?

This PR performs a much needed update to the main Campaign Action page gallery, as well as tackling a few bugs/issues that cropped up with the API and main DS app code (that also affect our products using the API) as the updates/refactoring of the gallery were taking place.
#### Where should the reviewer start?

Since there was a lot that needed to happen in this PR to switch over the gallery, I attempted to break the process down into [logical commits](https://github.com/DoSomething/phoenix/pull/5983/commits), so it might be best to go by each individual commit instead of checking all the files at once! Baby steps :baby: 
#### How should this be manually tested?

Pull down the branch, `drush cc all` and load up a gallery in a campaign action page. You can also do a few hits to the `/reportback-items` endpoint and make sure you get no errors.
#### Next steps
- More ideal if when caching the total tally for items based on query, we could specify the expiration time for the cached data longer than the default of 24 hours; would need to update ApiCache class to allow passing an expiration time in hours via `set()` method. I refrained for now because there is no way to clear that cache currently... which brings me to...
- On reportback save, update & delete, need to add a call to clear cache for RB item tally with specific campaign id (something like `cache_clear_all('reportback-items-tally_api_request|campaigns=:id', 'cache_dosomething_api', TRUE)`
- Remove **all** the old `gallery` API endpoint code for the old gallery!

---

@angaither @DFurnes 

CC: @mikefantini
